internal/azure/exchanger.go

```diff
@@ -0,0 +1,67 @@
+package azure
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+)
+
+type tokenResponse struct {
+	AccessToken  string `json:"access_token"`
+	RefreshToken string `json:"refresh_token"`
+	Resource     string `json:"resource"`
+	TokenType    string `json:"token_type"`
+}
+
+type Exchanger struct {
+	acrFQDN string
+}
+
+func NewExchanger(acrEndpoint string) *Exchanger {
+	return &Exchanger{
+		acrFQDN: acrEndpoint,
+	}
+}
+
+func (e *Exchanger) ExchangeACRAccessToken(armToken string) (string, error) {
+	exchangeUrl := fmt.Sprintf("%s://%s/oauth2/exchange", "https", e.acrFQDN)
+	parsedURL, err := url.Parse(exchangeUrl)
+	if err != nil {
+		return "", err
+	}
+
+	parameters := url.Values{}
+	parameters.Add("grant_type", "access_token")
+	parameters.Add("service", parsedURL.Hostname())
+	parameters.Add("access_token", armToken)
+
+	resp, err := http.PostForm(exchangeUrl, parameters)
+	if err != nil {
+		return "", fmt.Errorf("failed to send token exchnage request: %w", err)
+	}
+
+	if resp.StatusCode != 200 {
+		responseBytes, err := ioutil.ReadAll(resp.Body)
```

Comment:
Might be a good idea to not use `ioutil.ReadAll` here, we don't know how large the response body can be and this can cause memory issues. See my comment below.

Do we know what type of responses we get? If it's JSON maybe we can extract only the relevant bits for the error message. If not, maybe try to only read at a maximum just a few bytes worth of data.